{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cff5d9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pymol import cmd\n",
    "from Bio import SeqIO\n",
    "from Bio import Align,AlignIO\n",
    "from Bio.PDB import PDBParser\n",
    "from Bio.PDB.Polypeptide import standard_aa_names\n",
    "import re\n",
    "from modeller import *              # Load standard Modeller classes\n",
    "from modeller.automodel import *    # Load the AutoModel class\n",
    "from modeller.scripts import complete_pdb\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8603b6fb",
   "metadata": {},
   "source": [
    "In proteinfolder, the original pdb file and fasta file from protein data bank are required before running the following code\n",
    "the pdb file has name xxxx.pdb (e.g., 1t5z.pdb)\n",
    "the fasta file has name xxxx_1.fasta (e.g., 1t5z_1.fasta) (assuming the protein sequence is always 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9907d83",
   "metadata": {},
   "outputs": [],
   "source": [
    "proteinfolder = Path(Your_proteinfolder)\n",
    "proteinlist = [A List of PDB ID] #(e.g., ['2PIP','2PIO'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76c4a795",
   "metadata": {},
   "outputs": [],
   "source": [
    "def combine_fastanpdb(proteinid):\n",
    "    seqactual = ''\n",
    "    foundname = ''\n",
    "    for record in SeqIO.parse(proteinfolder/f'{proteinid}.pdb', \"pdb-atom\"):\n",
    "        name = re.sub('[^a-zA-Z0-9]', '_', record.id)\n",
    "        if len(str(record.seq)) > len(seqactual):\n",
    "            seqactual = str(record.seq)\n",
    "            seqactual = seqactual.replace('X','')\n",
    "            foundname = name\n",
    "\n",
    "    records = SeqIO.parse(proteinfolder/f\"{proteinid.upper()}_1.fasta\", \"fasta\")\n",
    "    for record in records:\n",
    "        nameref = foundname+'ref'\n",
    "        seqref = str(record.seq)\n",
    "\n",
    "    pdbid = foundname.split('_')[0]\n",
    "    with open(proteinfolder/f'{pdbid}.fasta', 'w') as f:\n",
    "        f.write(f\"> {pdbid} \\n\")\n",
    "        f.write(seqref+'\\n')    \n",
    "        f.write(f\"> {foundname} \\n\")\n",
    "        f.write(seqactual+'\\n')\n",
    "    f.close()\n",
    "    return foundname,seqref,seqactual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9553f3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def align2pir(name,seqref,seqactual):\n",
    "    pdbid = name.split('_')[0]\n",
    "    aligner = Align.PairwiseAligner()\n",
    "    alignments = aligner.align(seqref, seqactual)\n",
    "    aligresults = format(alignments[0],'fasta')\n",
    "    aligresults = aligresults.split('\\n')\n",
    "    aligresults[0] = f'> {pdbid}'\n",
    "    aligresults[2] = f'> {name}'\n",
    "    aligresults = [i+'\\n' for i in aligresults]\n",
    "    with open(proteinfolder/f'{pdbid}.fasta', 'w') as f:\n",
    "        f.writelines(aligresults)\n",
    "    f.close()\n",
    "\n",
    "    records = SeqIO.parse(proteinfolder/f\"{pdbid}.fasta\", \"fasta\")\n",
    "    count = SeqIO.write(records, proteinfolder/f\"{pdbid}.pir\", \"pir\")\n",
    "    alig_coord = alignments[0].coordinates\n",
    "    return alig_coord"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0746f264",
   "metadata": {},
   "outputs": [],
   "source": [
    "def pir2modellerpir(name,alig_coord):\n",
    "    pdbid = name.split('_')[0]\n",
    "    with open(proteinfolder/f'{pdbid}.pir', 'r') as pir:\n",
    "        pirlines = pir.readlines()\n",
    "    pir.close() \n",
    "\n",
    "    parser = PDBParser()\n",
    "    structure = parser.get_structure(f'{pdbid.lower()}', proteinfolder/f'{pdbid.lower()}.pdb')\n",
    "    structureinfo = {}\n",
    "    sequencelength = 0\n",
    "    for model in structure:\n",
    "        for chain in model:\n",
    "            protein_residues = [res for res in chain if res.get_resname() in standard_aa_names]\n",
    "            head_res = protein_residues[0].get_id()[1]\n",
    "            tail_res = protein_residues[-1].get_id()[1]\n",
    "            if sequencelength < (tail_res-head_res):\n",
    "                sequencelength = tail_res-head_res\n",
    "                structureinfo['chainID'] = chain.id\n",
    "                structureinfo['headres'] = head_res\n",
    "                structureinfo['tailres'] = tail_res\n",
    "    sequencehead = structureinfo['headres'] - (alig_coord[0][1]-alig_coord[1][1])\n",
    "    sequencetail = structureinfo['tailres'] + (alig_coord[0][-1]-alig_coord[1][-1])\n",
    "    namelist = []\n",
    "    headerlist = []\n",
    "    for index,line in enumerate(pirlines):\n",
    "        if '>' in line:\n",
    "            namelist.append(line.split(';')[1].replace('\\n',''))\n",
    "            pirlines[index] = pirlines[index].replace('XX','P1')\n",
    "            headerlist.append(index+1)\n",
    "    for name,index in zip(namelist,headerlist):\n",
    "        if '_' in name:\n",
    "            field1 = 'structureX'\n",
    "            field3 = structureinfo['headres']\n",
    "            field4 = structureinfo['chainID']\n",
    "            field5 = structureinfo['tailres']\n",
    "            field6 = structureinfo['chainID']\n",
    "        else:\n",
    "            field1 = 'sequence'\n",
    "            field3 = sequencehead\n",
    "            field4 = structureinfo['chainID']\n",
    "            field5 = sequencetail\n",
    "            field6 = structureinfo['chainID']\n",
    "        field2 = name\n",
    "        field7 = ' '\n",
    "        field8 = ' '\n",
    "        field9 = ' '\n",
    "        field10 = ' '\n",
    "        fieldlist = [field1,field2,field3,field4,field5,field6,field7,field8,field9,field10]\n",
    "        fieldlist = [str(i) for i in fieldlist]\n",
    "        newline = (':').join(fieldlist)\n",
    "        pirlines[index] = newline+'\\n'\n",
    "    with open(proteinfolder/f'{pdbid}.ali', 'w') as f:\n",
    "        f.writelines(pirlines)\n",
    "    f.close()    \n",
    "    shutil.copyfile(proteinfolder/f'{pdbid.lower()}.pdb',proteinfolder/f'{name}.pdb')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13f1de8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def modeller_exe(name):\n",
    "    pdbid = name.split('_')[0]\n",
    "    log.verbose()    # request verbose output\n",
    "    env = Environ()  # create a new MODELLER environment to build this model in    \n",
    "    \n",
    "    env.io.atom_files_directory = [str(proteinfolder)]\n",
    "\n",
    "    a = AutoModel(env,\n",
    "                alnfile  = proteinfolder/f\"{pdbid}.ali\",     # alignment filename\n",
    "                knowns   = f'{name}',              # codes of the templates\n",
    "                sequence = f'{pdbid}',              # code of the target\n",
    "\n",
    "    )\n",
    "    a.starting_model= 1                 # index of the first model\n",
    "    a.ending_model  = 5                 # index of the last model\n",
    "                                        # (determines how many models to calculate)\n",
    "    a.make()                                            # do comparative modeling  \n",
    "    scorelist = []\n",
    "    for i in range(1,6,1):\n",
    "        mdl = complete_pdb(env, f'{pdbid}.B9999000{str(i)}.pdb')\n",
    "\n",
    "        # Assess all atoms with DOPE:\n",
    "        s = Selection(mdl)\n",
    "        score = s.assess_dope()\n",
    "        scorelist.append(score)\n",
    "    refpdb = proteinfolder/f\"{pdbid.lower()}.pdb\"\n",
    "    current_directory = os.getcwd()\n",
    "    current_directory = current_directory.replace(\"\\\\\",\"/\")\n",
    "    selectstructure = str(1+np.argmin(scorelist))\n",
    "    shutil.move(f\"{current_directory}/{pdbid}.B9999000{selectstructure}.pdb\",proteinfolder/f\"{pdbid}.B9999000{selectstructure}.pdb\")\n",
    "    bestpdb = proteinfolder/f\"{pdbid}.B9999000{selectstructure}.pdb\"\n",
    "    cmd.load(refpdb)\n",
    "    cmd.load(bestpdb)\n",
    "    cmd.align(f'{pdbid}.B9999000{selectstructure}',f'{pdbid.lower()}')\n",
    "    cmd.save(proteinfolder/f\"{name.replace('_','')}.pdb\",selection=f'{pdbid}.B9999000{selectstructure}',format='pdb')\n",
    "    cmd.delete('all')  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89e7272e",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "for pro in proteinlist:\n",
    "    proteinid = pro.lower()\n",
    "    name,seqref,seqactual = combine_fastanpdb(proteinid)\n",
    "    if not os.path.exists(proteinfolder/f\"{name.replace('_','')}.pdb\"):\n",
    "        alig_coord = align2pir(name,seqref,seqactual)\n",
    "        pir2modellerpir(name,alig_coord)\n",
    "        modeller_exe(name)    \n",
    "        print(f'{pro} is done')"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
